containers/src/Data/Tree.hs

```diff
@@ -181,14 +182,43 @@ mfixTree f
 instance Traversable Tree where
     traverse f (Node x ts) = liftA2 Node (f x) (traverse (traverse f) ts)
 
+-- | Folds in preorder
 instance Foldable Tree where
-    foldMap f (Node x ts) = f x `mappend` foldMap (foldMap f) ts
+    fold = foldMap id
+    {-# INLINABLE fold #-}
+
+    foldMap f = go
+      where go (Node x ts) = f x `mappend` foldMap go ts
+    {-# INLINE foldMap #-}
+
+    foldr f z = \t -> go t z
+      where go (Node x ts) = f x . foldr ((.) . go) id ts
```

Comment:
Is there a reason not to tuck `z` under the lambda?

I imagine you pass `z` through the recursion because that performs better than doing the obvious thing. Have you verified that it's better? If so, a note explaining that would be helpful.

Reply:
I assume you mean "under the `go`"? Yes I had tried a few variations and found it performs well.
Added a comment.

Looking at the core and it does seem to generate better output. I wouldn't be able to explain why though.
For the record, on the `(&&)` benchmark
```hs
go (Node x ts) = f x . foldr ((.) . go) id ts
```
generates core like
```
Rec {
-- RHS size: {terms: 20, types: 19, coercions: 0, joins: 0/0}
go1 :: [Tree Bool] -> Bool -> Bool
[GblId, Arity=2, Str=<1L><ML>, Unf=OtherCon []]
go1
  = \ (ds :: [Tree Bool]) (eta :: Bool) ->
      case ds of {
        [] -> eta;
        : y ys ->
          case y of { Node ww1 ww2 ->
          case ww1 of {
            False -> False;
            True -> go1 ww2 (go1 ys eta)
          }
          }
      }
end Rec }
```
which is ideal, and
```hs
go (Node x ts) z' = f x (foldr go z' ts)
```
generates
```
Rec {
-- RHS size: {terms: 25, types: 27, coercions: 0, joins: 0/1}
$wgo3 :: Bool -> [Tree Bool] -> Bool -> Bool
[GblId, Arity=3, Str=<1L><ML><L>, Unf=OtherCon []]
$wgo3
  = \ (ww :: Bool) (ww1 :: [Tree Bool]) (w :: Bool) ->
      case ww of {
        False -> False;
        True ->
          letrec {
            go1 [Occ=LoopBreaker, Dmd=SCS(L)] :: [Tree Bool] -> Bool
            [LclId, Arity=1, Str=<1L>, Unf=OtherCon []]
            go1
              = \ (ds :: [Tree Bool]) ->
                  case ds of {
                    [] -> w;
                    : y ys -> case y of { Node ww3 ww4 -> $wgo3 ww3 ww4 (go1 ys) }
                  }; } in
          go1 ww1
      }
end Rec }
```